### ISSUE TITLE
[11.0] base_rest: RFC: Improve request/response Schema description

### ISSUE THREAD

**Issue by lmignon** [OCA/rest-framework#2]
## Current context

In base_Rest, information exchanged by services is validated according to a Cerberus schema defined as dict.

```python
    address_request_schema = {
        'name': {'type': 'string', 'required': True, 'empty': False},
        'street': {'type': 'string'},
    }
```

Defining such schemas could become boring and lead to a hard to read/maintain result when the schema is complex. This kind of definition is also hard to extend, reuse etc...

## Proposal

After much reflection and discussion with @sbidoul and some research into the web, I think that I'll base the schema definition of our data transfer objects on the  [pydantic](https://pydantic-docs.helpmanual.io) library. This library comes with a clean and standard way to define the structure of data transfer objects. It also provides methods to serialize/deserialize defined data transfer object to/from json and to generate the jsonschema from the objet definition.

~~Odoo already provides a clean api to define models. This API provides a pythonic way to define fields and relation between models. It also provides a clean way to override and extend models across modules. The Idea is to define a new base class based on BaseModel to define the messages exchanged by the services.~~

```python

class RestModel(models.BaseModel):
    _name='rest.model'
    
    def from_json(self, json_data):
        return self.new(....)

    def to_json(self):
        return {}      
    
    def validate(self):
        return True

    def to_json_schema(self):
        """ Return the definition of the model as json schema
        see http://json-schema.org/
        """
        retrun {}


class AddressRequest(RestModel):
    _name = 'address.request'

    name = fields.Char(required=True)
    street = fields.Char()

    ....
```

~~Services will be annoted to specify the request/response models instead of the current implicit properties. This annotation will also be used to say that the request  / response is a list or not (default)~~ see #6

### Expected benefits:

* More modular
* Easier to extend / reuse / redefine
* Easier to manipulate into the services implementation. `my_address['name'] vs my_address.name`
* Easier to document (thanks to `string` and `help`attributes of the fields we'll be able to provide a description into the json schema)
* ....

**Comment by sbidoul** [comment]
:+1:

**Comment by lmignon** [comment]
@sebastienbeau @sbidoul 
I wonder if it makes sense to base the definition of messages exchanged on Odoo models. At this stage I have the feeling that the Odoo models are too complex objects that come with a lot of features specific to their use by the ORM. The'id' field for example is implicit in these models. We also have a strong management of the cache relative to an ORM without speaking of the advanced functionnalities allowing the definition of calculated fields etc....
Do we really need all these stuff in our 'data models/ json models'? I feel it would be safer to develop our own API in order to master the functional scope of these models and especially the way they are serialized/deserialized while having the possibility to extend them in the same way we extend an Odoo model.
What do you think about that? 

lmi

**Comment by sbidoul** [comment]
So you mean you'd use Odoo models to declare schemas but generate Plain Old Python Objects when deserializing?

**Comment by lmignon** [comment]
I mean I wonder if it would be easier to have our own base model (with the same extension mechanism that the one provided by '_inherit') that doesn't inherit BaseModel, and to have our own fields definitions.

**Comment by sbidoul** [comment]
Do we necessarily need a base model for the objects? They can be plain python objects.

If I understood correctly the main motivation to use something else than Cerberus is the expressiveness of the schema language (difficulty of reusing schema fragments, no help/doc, ...).

**Comment by farzeni** [comment]
Hello, 

did you considered an approach like the one of django rest framework? 
https://www.django-rest-framework.org/api-guide/serializers/

In that case Serializers are a decoupled class which takes care of the schema definition as well as performing base CRUD methods. This allows dev to use them with custom python objects and  with ORM objects (using specific ModelSerializer).

**Comment by lmignon** [comment]
@farzeni That's the way base_rest is implement (except schema) Services are components not linked to the ORM. The challenge here is to find a way to define input/output schemas that could be extended by an addon extending a base one. There are existing python libs (https://pydantic-docs.helpmanual.io/ , https://marshmallow.readthedocs.io/ ) providing a simple way to define these schemas but we lost the possibility to extend these schemas at runtime...

**Comment by lmignon** [comment]
This RFC is implemented by #41

**Comment by lmignon** [comment]
merged with #93 #48 and #71
